grype/db/v6/distribution/client.go

```diff
@@ -0,0 +1,234 @@
+package distribution
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"time"
+
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/spf13/afero"
+	"github.com/wagoodman/go-progress"
+
+	"github.com/anchore/clio"
+	v6 "github.com/anchore/grype/grype/db/v6"
+	"github.com/anchore/grype/internal/file"
+	"github.com/anchore/grype/internal/log"
+)
+
+type Config struct {
+	ID clio.Identification
+
+	// check/fetch parameters
+	LatestURL string
+	CACert    string
+
+	// validations
+	ValidateByHashOnGet bool
+	RequireUpdateCheck  bool
+
+	// timeouts
+	CheckTimeout  time.Duration
+	UpdateTimeout time.Duration
+}
+
+type Client interface {
+	IsUpdateAvailable(current *v6.Description) (*Archive, error)
+	Download(archive Archive, dest string, downloadProgress *progress.Manual) (string, error)
+}
+
+type client struct {
+	fs               afero.Fs
+	latestHTTPClient *http.Client
+	updateDownloader file.Getter
+	config           Config
+}
+
+func DefaultConfig() Config {
+	return Config{
+		LatestURL:           "https://grype.anchore.io/databases/latest.json",
+		ValidateByHashOnGet: true,
+		RequireUpdateCheck:  false,
+		CheckTimeout:        30 * time.Second,
+		UpdateTimeout:       300 * time.Second,
+	}
+}
+
+func NewClient(cfg Config) (Client, error) {
+	fs := afero.NewOsFs()
+	latestClient, err := defaultHTTPClient(fs, cfg.CACert, withClientTimeout(cfg.CheckTimeout))
+	if err != nil {
+		return client{}, err
+	}
+
+	dbClient, err := defaultHTTPClient(fs, cfg.CACert, withClientTimeout(cfg.UpdateTimeout))
+	if err != nil {
+		return client{}, err
+	}
+
+	return client{
+		fs:               fs,
+		latestHTTPClient: latestClient,
+		updateDownloader: file.NewGetter(cfg.ID, dbClient),
+		config:           cfg,
+	}, nil
+}
+
+// IsUpdateAvailable indicates if there is a new update available as a boolean, and returns the latest db information
+// available for this schema.
+func (c client) IsUpdateAvailable(current *v6.Description) (*Archive, error) {
+	log.Debugf("checking for available database updates")
+
+	updateEntry, err := c.latestFromURL()
+	if err != nil {
+		if c.config.RequireUpdateCheck {
+			return nil, fmt.Errorf("check for vulnerability database update failed: %+v", err)
+		}
+		log.Warnf("unable to check for vulnerability database update")
+		log.Debugf("check for vulnerability update failed: %+v", err)
+	}
+	return c.isUpdateAvailable(current, updateEntry)
+}
+
+func (c client) isUpdateAvailable(current *v6.Description, candidate *LatestDocument) (*Archive, error) {
+	if candidate == nil {
+		return nil, nil
+	}
+	log.Debugf("found database update candidate: %s", candidate.Archive.Description)
```

Comment:
This log message is a little confusing - it sounds like we've already decided to update. Maybe something like `"comparing current db against ..."` or just skip the log here since we'll always say the update is or isn't available.

Reply:
removed